presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java

```diff
@@ -390,8 +392,42 @@ else if (firstDependencies.stream().allMatch(right::canResolve) && secondDepende
                 }
             }
 
-            // subqueries can be applied only to one side of join - left side is selected in arbitrary way
-            leftPlanBuilder = subqueryPlanner.handleUncorrelatedSubqueries(leftPlanBuilder, complexJoinExpressions, node, context);
+            if (node.getType() == LEFT || node.getType() == RIGHT) {
+                RelationType left = analysis.getOutputDescriptor(node.getLeft());
+                RelationType right = analysis.getOutputDescriptor(node.getRight());
+
+                for (Expression complexJoinExpression : complexJoinExpressions) {
+                    // IN subqueries are already disallowed previously
+                    if (subqueryPlanner.collectScalarSubqueries(complexJoinExpression, node).isEmpty() &&
+                            subqueryPlanner.collectExistsSubqueries(complexJoinExpression, node).isEmpty() &&
+                            subqueryPlanner.collectQuantifiedComparisonSubqueries(complexJoinExpression, node).isEmpty()) {
+                        continue;
+                    }
+
+                    Set<QualifiedName> dependencies = VariablesExtractor.extractNames(complexJoinExpression, analysis.getColumnReferences());
+                    if (dependencies.isEmpty() || (dependencies.stream().anyMatch(left::canResolve) && dependencies.stream().anyMatch(right::canResolve))) {
+                        // Fallback to legacy behavior
+                        // subqueries is applied only to one side of join - left side is selected in arbitrary way
+                        leftPlanBuilder = subqueryPlanner.handleUncorrelatedSubqueries(leftPlanBuilder, ImmutableList.of(complexJoinExpression), node, context);
+                    }
+                    else {
+                        if (node.getType() == LEFT && !dependencies.stream().anyMatch(left::canResolve)) {
+                            rightPlanBuilder = subqueryPlanner.handleSubqueries(rightPlanBuilder, complexJoinExpression, node, context);
+                        }
+                        else if (node.getType() == RIGHT && !dependencies.stream().anyMatch(right::canResolve)) {
+                            leftPlanBuilder = subqueryPlanner.handleSubqueries(leftPlanBuilder, complexJoinExpression, node, context);
+                        }
+                        else {
+                            leftPlanBuilder = subqueryPlanner.handleSubqueries(leftPlanBuilder, ImmutableList.of(complexJoinExpression), node, context);
```

Comment:
1. why are you wrapping the complexJoinExpression in a list here?  
2. i was assuming before that we are only pushing correlations to the inside of an outer join (and that's also the same reason we exclude full join.  wouldn't this mean for left joins this would need to be handleUncorrelatedSubqueries?  
3. why is there no case for pushing to the right side of right joins?

Reply:
> 1. why are you wrapping the complexJoinExpression in a list here?
I can use the other function - will fix

> 2. i was assuming before that we are only pushing correlations to the inside of an outer join (and that's also the same reason we exclude full join.  wouldn't this mean for left joins this would need to be handleUncorrelatedSubqueries?
> 3. why is there no case for pushing to the right side of right joins?
We may be using different nomenclature here. When I say  inner side of outer joins, I mean the non-preserving (build) side. Just as predicates on the outer (preserving) side of outer joins may not be pushed below the join, correlated subqueries cannot be pushed down as well